feat(xo-web,xo-server): ability to import a VHD/VMDK disk

nraynaud opened the pull request:
Fixes  #4137

- The demo route is /import-vmdk
- the file xo-web/src/xo/index.js needs a translation
- There is a confirmation message in the browser console after import is done.

@Rajaa-BARHTAOUI here is the thing.

### Check list

> Check items when done or if not relevant

- [ ] PR reference the relevant issue (e.g. `Fixes #007`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] `CHANGELOG.unreleased.md`:
   - enhancement/bug fix entry added
   - list of packages to release updated (`${name} v${new version}`)
- [ ] documentation updated
- [ ] **I have tested added/updated features** (and impacted code)

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer
1. if necessary, update your PR, and re- add a reviewer

nraynaud committed:
VMDK import prototype.
 - The demo route is /import-vmdk
 - the file xo-web/src/xo/index.js needs a translation
 - There is a confirmation message in the browser console after import is done.

nraynaud committed:
VMDK import prototype - fix table extraction

nraynaud commented:
Can you review only packages/xo-server and packages/xo-vmdk-to-vhd please?

nraynaud committed:
VMDK import prototype - add VHD import

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +167,53 @@ resize.params = {
+  { capacity, name, blocksTable, srId, xapi }
+) {
+  // Timeout seems to be broken in Node 4.
+  // See https://github.com/nodejs/node/issues/3319
```
You can remove this comment, Node 4 is not supported by XO.

nraynaud committed:
VMDK import prototype - remove deprecated comment

julien-f commented on packages/vhd-lib/src/peekFooterFromVhdStream.js:
```diff
@@ -0,0 +1,10 @@
+  const footer = fuFooter.unpack(footerBuffer)
+  stream.unshift(footerBuffer)
+  return footer
+}
```
Don't forget to add `vhd-lib` to the changelog or we'll forget to release it.

julien-f committed:
disk.import

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,72 @@ resize.params = {
+    })
+    req.length = req.headers['content-length']
+
+    if (vhdStream === undefined)
```
Test before creating the VDI.

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,72 @@ resize.params = {
+    if (vhdStream === undefined)
+      throw new Error(
+        `Unknown disk type, expected "vhd" or "vmdk", got ${type}`
+      )
```
Always add braces to conditional/loop statements.

julien-f commented on packages/xo-vmdk-to-vhd/src/index.js:
```diff
@@ -14,4 +16,8 @@ async function convertFromVMDK(vmdkReadStream, table) {
+  convertFromVMDK as default,
+  readVmdkGrainTable,
+  readCapacityAndGrainTable,
+}
```
Add to changelog or it won't be released.

nraynaud committed:
VMDK import prototype - update changelog

nraynaud committed:
VMDK import prototype - rename disk.import

nraynaud committed:
VMDK import prototype - fix control flow

julien-f commented on CHANGELOG.md:
```diff
@@ -1,5 +1,31 @@
+- vhd-lib v0.6.1
+- xo-server v5.39.0
+- xo-web v5.39.1
+
```
Please, revert these changes.

Keep me if you need help next time rebasing.

julien-f commented on CHANGELOG.unreleased.md:
```diff
@@ -1,23 +1,12 @@
-- xo-vmdk-to-vhd v0.1.7
-- vhd-lib v0.6.1
-- xo-server v5.39.0
-- xo-web v5.39.0
```
Please revert.

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,72 @@ resize.params = {
+      vhdStream = await convertFromVMDK(req, vmdkData.blocksTable)
+      size = vmdkData.size
+    }
+    if (type === 'vhd') {
```
`else if`

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,72 @@ resize.params = {
+      const footer = await peekFooterFromVhdStream(req)
+      size = footer.currentSize
+    }
+    if (vhdStream === undefined) {
```
`} else {`

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,72 @@ resize.params = {
+    const vdi = await xapi.createVdi({
+      name_label: name,
+      size,
+      sr: srId && this.getObject(srId).$ref,
```
Is `srId` optional?

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,72 @@ resize.params = {
+      name,
+      vmdkData,
+      srId: sr._xapiId,
+      xapi: this.getXapi(sr),
```
Why storing the `xapi` object and not the `sr`?

julien-f replied:
?

julien-f replied:
Still no answer on this.

nraynaud replied:
sorry, missed it. will do.

julien-f replied:
It was not a request, just a question :slightly_smiling_face:

nraynaud replied:
It's a copy/paste of the code in the OVA import, I did not write the original.

I did a copy/paste in the hope that it would go through faster, but that did not happen.

julien-f replied:
Ok :stuck_out_tongue:

julien-f commented on packages/xo-server/src/xo-mixins/jobs/index.js:
```diff
@@ -266,6 +266,11 @@ export default class Jobs {
+      data = {
+        reportWhen: job.settings['']?.reportWhen ?? 'failure',
+      }
+    }
```
Unrelated change, something went wrong with your rebase.

julien-f commented on packages/xo-vmdk-to-vhd/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "xo-vmdk-to-vhd",
-  "version": "0.1.6",
+  "version": "0.1.7",
```
Don't change version numbers except in the changelog, it's done during the release process.

julien-f committed:
chore: extract getLongLong

julien-f committed:
chore: replace reduce with every

julien-f commented on packages/xo-vmdk-to-vhd/src/vmdk-read-table.js:
```diff
@@ -51,47 +60,52 @@ export default async function readVmdkGrainTable(fileAccessor) {
-  )
-  return extractedGrainTable.map(([index, grainAddress]) => index * grainSize)
+
+  return { tablePromise: readTable(), capacityBytes: capacity }
```
Why did you choose to return a promise?

nraynaud replied:
the table extraction can be a long process, that way it start immediately upon dropping the file on the UI, but it only becomes blocking when we submit the file.

julien-f replied:
This is weird, but as long as it works.

julien-f committed:
chore: various style changes

nraynaud committed:
VMDK import prototype - fix control flow

nraynaud commented:
it should be your last review for this part

julien-f commented on CHANGELOG.unreleased.md:
```diff
@@ -12,3 +12,5 @@
 - xo-server v5.40.0
 - xo-web v5.40.0
+- vhd-lib v0.6.2
+- xo-vmdk-to-vhd v0.1.8
```
Packages are released in order, please keep dependencies above.

nraynaud replied:
wait, I'm pretty sure I got scolded for putting stuff above before.

julien-f replied:
I cannot find such a comment but maybe it was the order in between your packages that was incorrect.

nraynaud committed:
VMDK import prototype - pass SR directly

pdonias commented:
Should we merge the UI branch into this branch? @nraynaud?

nraynaud commented:
Yes I think it’s a good idea.

Envoyé de mon iPhone

> Le 30 avr. 2019 à 01:26, Pierre Donias <notifications@github.com> a écrit :
> 
> Should we merge the UI branch into this branch? @nraynaud?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

rajaa212 committed:
Import disk UI (#4142)

pdonias committed:
Fix error typo

pdonias committed:
Fix srId

pdonias committed:
Add loading spinner

nraynaud committed:
VMDK import prototype - fix VMDK import

olivierlambert reviewed (approved):
Tested with VMDK and VHD files: OK

julien-f committed:
fix: use JsonRpcError

olivierlambert commented:
@julien-f what's left to merge this?

julien-f commented:
@olivierlambert Nothing, I just need to finish reviewing this and merging this, it's on my PRs list :slightly_smiling_face:

olivierlambert commented:
Roger that!

julien-f commented on packages/xo-server/src/api/disk.js:
```diff
@@ -165,3 +168,77 @@ resize.params = {
+      size,
+      sr: srId,
+    })
+    await xapi.importVdiContent(vdi, vhdStream, VDI_FORMAT_VHD)
```
Should we destroy the VDI if this fails?

julien-f commented:
In the first comment:
- what is the demo route?
- is a translation still needed?

Also, please check the relevant boxes in the todo list.

julien-f committed:
various updates

julien-f commented:
@nraynaud It appears we forgot this PR, can you please take a quick look and tell me if it's safe to merge for this release?

nraynaud commented:
Sorry, it doesn't work on the first try, and I see some suspicious integer code. It needs debugging.

julien-f commented:
Ok, I'm marking this as WiP and we'll merge this later.

nraynaud committed:
fix execa use in tests.

nraynaud commented:
I am making a big change, I think we'll need to re-do the review cycle.

julien-f commented:
No problem, keep me posted :slightly_smiling_face: 

And don't forget to rebase your branch on top of `master`.

nraynaud committed:
read and retrieve the grain table on the browser and use it server side
 - this allows for reading both streamOptimized and normal VMDK

nraynaud committed:
fix changelog

nraynaud commented:
Sorry, I made a big change, we might have to go through the entire review rigamarole again.

nraynaud commented:
Now the strategy is to always read the grain table, and use that to determine where are the grains, whereas before, we would read the markers to find the grains in the sequential stream of streamOptimized files. Of course, some people would feed it non-streamOptimized files and nothing would work. So now I always read the grain table and use it server side to read the grains.

nraynaud committed:
more consistent naming
 blockAddressList -> grainLogicalAddressList
 grainsAddressList -> grainFileOffsetList

nraynaud committed:
more consistent naming
 currentBlock -> currentFragmentIndex

nraynaud committed:
on upload error delete the newly created VDI

nraynaud committed:
fix rebase issue

julien-f commented on packages/vhd-lib/package.json:
```diff
@@ -22,12 +22,13 @@
   "dependencies": {
     "@xen-orchestra/log": "^0.2.0",
-    "async-iterator-to-stream": "^1.0.2",
+    "async-iterator-to-stream": "^1.0.3",
```
Why?

And why not update it to the latest version?

nraynaud committed:
upgrade async-iterator-to-stream to last version

nraynaud commented:
STOP THE PRESS! I broke the .ova import, on it :(

nraynaud committed:
fix .ova vm import.

nraynaud commented:
Should be fine by now.

nraynaud committed:
improve VMDK notes

nraynaud committed:
improve VMDK notes

julien-f commented on packages/vhd-lib/package.json:
```diff
@@ -22,12 +22,13 @@
   "dependencies": {
     "@xen-orchestra/log": "^0.2.0",
-    "async-iterator-to-stream": "^1.0.2",
+    "async-iterator-to-stream": "^1.1.0",
```
I will revert this because not necessary for this PR.

nraynaud committed:
code style improvement

julien-f committed:
revert unnecessary changes

julien-f commented on packages/vhd-lib/src/createReadableSparseStream.js:
```diff
@@ -17,38 +18,65 @@ import { set as setBitmap } from './_bitmap'
+    if (!lastFragmentPerBlock.has(vhdTableIndex)) {
+      lastFragmentPerBlock.set(vhdTableIndex, fragmentLogicAddress)
     }
   })
```
Wouldn't it be easier to iterate forward and to always assign the new address?

julien-f replied:
This may also remove the need of `forEachRight` for the next iteration.

nraynaud replied:
I don't think so, because the VMDK grain representing a VHD block might be interleaved.
let's say the VHD blocks `A` and `B` are fragmented in 3 VMDK grains each and that the VMDK stream present the grains in this order (letters are VHD, numbers are VMDK):
`A3 B1 A1 B2 A2`

The VHD BAT using the least intermediate memory possible is `B A` because I can only write the block A on the stream when I receive `A1`, and it occurs after the last Block of `B`. Note that B3 doesn't even appears in the VMDK, its bits in the VHD block bitmap will just stay at zero, and the VHD block will be dumped on the VHD stream when encountering `B2` because we know it's the last time we see a block from `B`.

nraynaud committed:
describe conversion algorithm in notes

julien-f committed:
format with Prettier

julien-f committed:
vhd-lib needs to be released
